toolchain: iar: Experimental IAR support

RobinKastberg opened the pull request:
This PR integrates experimental IAR Toolchain support into Zephyr.
Currently supporting selected ARM Cortex-M targets.
We have run internal testing where:

The following have at various points passed `--level acceptance` (and much more) using the IAR Toolchain

* `ek_ra4e2` (currently misses some CMAKE_LINKER_GENERATOR sections)
* `frdm_mcxn947/mcxn947/cpu0`
* `mimxrt1060_evk`
* `nrf52840dk/nrf52840`
* `qemu_cortex_m0`
* `qemu_cortex_m3`

Currently this PR is being tested using `--level acceptance` on `qemu_cortex_m3`

## Limitations/Known Issues
* Currently only minimallibc is supported, this means Picolibc and Newlib is not supported. There is experimental support for IARs DLib.
* Currently C++ is not supported
* Currently TrustZone is not working. 
* Currently `CONFIG_USERSPACE` is not working due to the usage of `CMAKE_LINKER_GENERATOR`. Support for `CONFIG_USERSPACE` is being worked on internally in IAR.
* `--no_system_include` should probably be used but isn't currently. We are unsure if this will cause problems.
* Currently using the GNU Assembler for .S files
* ~Currently `CONFIG_STACK_CANARIES` is not working due to a different type of `__stack_chk_guard`~
* ~The current method for static initialization in the linker script generator is experimental and will most likely change during this PR or shortly after this PR.~

## Obtaining an IAR Toolchain

Right now, a special development version of the IAR build tools for Arm is required to work with Zephyr. 
Toolchain can be found here: https://github.com/iarsystems/zephyr-iar/releases

A license is required to use this toolchain.
It is distributed to selected partners and customers for evaluation. If you are interested in being 
part of this program, please send a request to our FAE team at fae.emea@iar.com.

## TODO
 - [x] Copyright checks
 - [x] Compliance checks
 - [x] https://github.com/zephyrproject-rtos/zephyr/pull/84654 (currently included in PR, but can and should be merged separately and removed from here) 
 - [x] https://github.com/zephyrproject-rtos/zephyr/pull/84481 (currently included in PR, but needs to be merged separately) 
 - [x] https://github.com/zephyrproject-rtos/cmsis/pull/25 (module PR for above) 
 - [x] https://github.com/zephyrproject-rtos/zephyr/pull/85052 (currently included in this PR to make tests gcc x86 tests pass, should be merged separately and removed from here)
 - [x] https://github.com/zephyrproject-rtos/zephyr/pull/84862 (workaround currently included in this PR to make IAR tests pass)

nordicjm commented on cmake/compiler/iar/compiler_flags.cmake:
```diff
@@ -0,0 +1,167 @@
+
+set_compiler_property(PROPERTY optimization_debug -Ol)
+
+set_compiler_property(PROPERTY optimization_speed -Ohs)
```
there is also `optimization_size_aggressive`

danielfl-iar replied:
iirc we don't have anything more aggressive than `Ohz`

LoveKarlsson replied:
We only have a speed aggressive option

nordicjm commented on cmake/compiler/iar/compiler_flags.cmake:
```diff
@@ -0,0 +1,167 @@
+  --diag_suppress=Pe513  # a value of type "void *" cannot be assigned to an entity of type "int (*)(int)"
+)
+
+set(IAR_WARNING_DW3 )
```
```suggestion
set(IAR_WARNING_DW3)
```

nordicjm commented on cmake/compiler/iar/compiler_flags.cmake:
```diff
@@ -0,0 +1,167 @@
+
+set(IAR_WARNING_DW3 )
+
+set_compiler_property(PROPERTY warning_dw_1
+  ${IAR_WARNING_DW_3}
+  ${IAR_WARNING_DW_2}
+  ${IAR_WARNING_DW_1} )
```
no space at end and stick to one style, line 42 has ) on newline without indent, line 23 has it on newline with indent, update all to be the same

nordicjm commented on cmake/compiler/iar/compiler_flags.cmake:
```diff
@@ -0,0 +1,167 @@
+# Flag for disabling rtti in C++
+set_property(TARGET compiler-cpp PROPERTY no_rtti --no_rtti)
+
+
```
```suggestion
```

nordicjm commented on cmake/compiler/iar/generic.cmake:
```diff
@@ -0,0 +1,13 @@
+  message(FATAL_ERROR "Zephyr was unable to find \`arm-zephyr-eabi-gcc\` for DTS preprocessing")
+endif()
+
+if (CMAKE_C_COMPILER STREQUAL CMAKE_C_COMPILER-NOTFOUND)
```
no space between if and bracket in cmake

nordicjm commented on cmake/compiler/iar/iccarm-cpu.cmake:
```diff
@@ -0,0 +1,88 @@
+# KConfig'uration and sets this to ICCARM_CPU
+
+if("${ARCH}" STREQUAL "arm")
+  if    (CONFIG_CPU_CORTEX_M0)
```
spacing to go

nordicjm commented on cmake/compiler/iar/iccarm-cpu.cmake:
```diff
@@ -0,0 +1,88 @@
+    if    (CONFIG_ARMV8_1_M_MVEF)
+      set(ICCARM_CPU Cortex-M55)
+    elseif(CONFIG_ARMV8_1_M_MVEI)
+      # set(ICCARM_CPU cortex-m55.no_mve.fp)
```
commented line can go?

nordicjm commented on cmake/compiler/iar/iccarm-fpu.cmake:
```diff
@@ -0,0 +1,52 @@
+# KConfig'uration and sets this to ICCARM_FPU
+
+if(CONFIG_FPU)
+
```
```suggestion
```
lines below missing indent

nordicjm replied:
not fixed

nordicjm commented on cmake/compiler/iar/iccarm-fpu.cmake:
```diff
@@ -0,0 +1,52 @@
+    set(FPU_FOR_Cortex-M33          FPv5${PRECISION_TOKEN})
+    set(FPU_FOR_Cortex-M33.no_dsp   FPv5${PRECISION_TOKEN})
+    set(FPU_FOR_Cortex-M55          auto)
+    #set(FPU_FOR_cortex-m55+nomve.fp auto)
```
as above

nordicjm commented on cmake/compiler/iar/iccriscv-cpu.cmake:
```diff
@@ -0,0 +1,100 @@
+endif()
+
+if (CONFIG_RISCV_ISA_RV32E)
+    string(CONCAT riscv_mabi ${riscv_mabi} "e")
```
cmake indent is 2 spaces

nordicjm commented on cmake/compiler/iar/target.cmake:
```diff
@@ -0,0 +1,162 @@
+  list(APPEND IAR_COMMON_FLAGS
+    --endian=little
+    --cpu=${ICCARM_CPU}
+
```
empty lines can go

nordicjm commented on cmake/compiler/iar/target.cmake:
```diff
@@ -0,0 +1,162 @@
+    --diag_suppress=Ta184  # Using zero sized arrays except for as last member of a struct is discouraged and dereferencing elements in such an array has undefined behavior
+  )
+else()
+  list(APPEND IAR_COMMON_FLAGS
+
+  )
```
???

nordicjm commented on cmake/toolchain/iar/Kconfig:
```diff
@@ -0,0 +1,43 @@
+config IAR_SEMIHOSTING
+	bool "Use the IAR semihosting implementation."
+	depends on IAR_LIBC
+	default n
```
`default n` is not needed

nordicjm commented on cmake/toolchain/iar/Kconfig:
```diff
@@ -0,0 +1,43 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config LD_LINKER_SCRIPT_SUPPORTED
+	bool
```
do not redeclare existing Kconfigs

RobinKastberg replied:
This is how armclang does it, which we were shamelessly inspired by. 
What is the proper way to do it?

nordicjm replied:
drop the `bool` line, then it's OK

nordicjm commented on cmake/toolchain/iar/Kconfig.defconfig:
```diff
@@ -0,0 +1,70 @@
+	default y
+
+config PICOLIBC_SUPPORTED
+	bool
```
as above

nordicjm commented on include/zephyr/arch/arm/asm_inline.h:
```diff
@@ -14,10 +14,10 @@
+#if defined(__GNUC__) || defined(__ICCARM__)
 #include <zephyr/arch/arm/asm_inline_gcc.h>
 #else
-#include <arch/arm/asm_inline_other.h>
```
looks like you are breaking something here

RobinKastberg replied:
This file does not exist

nordicjm commented on scripts/ci/check_compliance.py:
```diff
@@ -474,6 +474,32 @@ def get_kconfig_dts(self, kconfig_dts_file, settings_file):
         except subprocess.CalledProcessError as ex:
             self.error(ex.output.decode("utf-8"))
 
+    def get_v1_model_syms(self, kconfig_v1_file, kconfig_v1_syms_file):
```
what is going on here?

RobinKastberg replied:
Looks like I have done something weird when doing my PR getting an old version of the file..
I will take a look at it.
What I was trying to do was adding some Kconfig symbols that were unknown.

nordicjm reviewed (commented):
had a very quick look at this and left some comments, can you apply the suggestions from comments to the whole PR (only some instances have been marked)?

RobinKastberg commented:
> had a very quick look at this and left some comments, can you apply the suggestions from comments to the whole PR (only some instances have been marked)?

Yes! Thanks! Some of those are in compliance checker as well. 
For some reason compliance checker locally doesn't complain the same and I don't understand why.

nashif commented:
lets try and get this into 4.1, even in the experimental state, this is a great addition and it is standalone enough.

RobinKastberg commented on cmake/compiler/gcc/compiler_flags.cmake:
```diff
@@ -231,8 +231,13 @@ set_property(TARGET compiler-cpp PROPERTY no_threadsafe_statics "-fno-threadsafe
 # gcc flag for colourful diagnostic messages
-check_set_compiler_property(PROPERTY diagnostic -fdiagnostics-color=always)
+if (NOT COMPILER STREQUAL "xcc")
+set_compiler_property(PROPERTY diagnostic -fdiagnostics-color=always)
```
this was accidental anyway, the intended changeset in this file is just adding the asm property.

RobinKastberg committed:
checkpatch: Allow __BYTE_ORDER__ and __aligned__ in toolchain files

Checkpatch needs to allow toolchain headers where
these macros are defined.

Signed-off-by: Robin Kastberg <robin.kastberg@iar.com>

RobinKastberg committed:
toolchain: move some hardcoded flags to properties

Some basic flags that were hardcoded needed
to be moved to properties to be able to implement
alternative toolchains.

Signed-off-by: Robin Kastberg <robin.kastberg@iar.com>

RobinKastberg committed:
toolchain: iar: Add experimental IAR support

This adds experimental support for the IAR toolchain.

Signed-off-by: Robin Kastberg <robin.kastberg@iar.com>

RobinKastberg committed:
toolchain: iar: tls: no tls pointer offset in IAR

The IAR Toolchain currently doesn't use the two
pointer offset.

Signed-off-by: Robin Kastberg <robin.kastberg@iar.com>

RobinKastberg committed:
MAINTAINERS: add the `Toolchain IAR` area

This commit adds the `Toolchain IAR` area of maintenance.

Signed-off-by: Robin Kastberg <robin.kastberg@iar.com>

RobinKastberg committed:
toolchain: doc: Add toolchain information about IAR

This commit adds documentation information
about the IAR toolchain.

Signed-off-by: Robin Kastberg <robin.kastberg@iar.com>

RobinKastberg committed:
toolchain: stop C flags from leaking to the assembler

Currently the compiler and assembler shares many
properties. This can be problematic for non-GNU
toolchains that takes different parameters to the
assembler and compiler.

Signed-off-by: Robin Kastberg <robin.kastberg@iar.com>

cfriedt reviewed (dismissed):
I hope there is some way we can get this into CI, just to reduce the possibility of regressions for IAR users.

Good work though 👍

RobinKastberg commented:
> I hope there is some way we can get this into CI, just to reduce the possibility of regressions for IAR users.
> 

There is nothing practical stopping us. It has been a goal of IAR all along. However we have identified the following issues:
* CI resources/pipeline runtime has been a worry of the community.
* Tuning warning levels and tests to avoid making PRs overly difficult. IAR toolchain is stricter in many ways, but hopefully this will be seen as increasing code quality and standards compliance.
* Some way for PR makers to be able to get a hold of our tools/licenses IF something goes wrong in CI, so they can test locally. I could imagine this being very frustrating.
* In practice including the toolchain zip in CI and adding a github secret with a license key. I don't know exactly how to make this happen, but I wouldn't imagine it being particularly difficult.

Currently we run in-house pipelines that will point to main as soon as we get merged, but these will by nature be reactive.
We are experimenting with running in-house jobs on PRs, but that is not being run consistently right now.

kartben commented:
@tejlmand could you provide some feedback on this one, and do you think it can make into 4.1? Thanks!

nordicjm commented on cmake/compiler/iar/iccarm-fpu.cmake:
```diff
@@ -0,0 +1,53 @@
+  # 64-bit
+  else()
+    set(ICCARM_FPU none)
+  endif()
+
```
```suggestion
```

nordicjm commented on cmake/compiler/iar/iccarm-fpu.cmake:
```diff
@@ -0,0 +1,53 @@
+      set(FPU_FOR_Cortex-M55.no_mve   auto)
+      # We don't have this one?
+      set(FPU_FOR_Cortex-M55.no_dsp   auto)
+
```
```suggestion
```

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+        EXPR "@ADDR(${symbol_start})@"
+        )
+    endif()
+
```
```suggestion
```

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+    create_symbol(OBJECT ${REGION_OBJECT} SYMBOL __${name}_size
+      EXPR "(@(__${name}_end)@ - @(__${name}_start)@)"
+      )
+
```
```suggestion
```
not going to mark more, apply changes throughout file

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+  endif()
+
+  foreach(region ${regions})
+    get_property(name    GLOBAL PROPERTY ${region}_NAME)
+    get_property(address GLOBAL PROPERTY ${region}_ADDRESS)
+    get_property(flags   GLOBAL PROPERTY ${region}_FLAGS)
+    get_property(size    GLOBAL PROPERTY ${region}_SIZE)
```
```suggestion
    get_property(name GLOBAL PROPERTY ${region}_NAME)
    get_property(address GLOBAL PROPERTY ${region}_ADDRESS)
    get_property(flags GLOBAL PROPERTY ${region}_FLAGS)
    get_property(size GLOBAL PROPERTY ${region}_SIZE)
```
also needs applying throughout

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+    set(${STRING_STRING} "${${STRING_STRING}}  define block HEAP with alignment=8, expanding size { };\n")
+    set(${STRING_STRING} "${${STRING_STRING}}}\n")
+    set(${STRING_STRING} "${${STRING_STRING}}\"DLib heap\": place in RAM { block HEAP };\n")
+#    set(${STRING_STRING} "${${STRING_STRING}}define exported symbol HEAP$$Base=kheap__system_heap;\n")
+#    set(${STRING_STRING} "${${STRING_STRING}}define exported symbol HEAP$$Limit=END(kheap__system_heap);\n")
```
to go?

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+        else()
+          set(${STRING_STRING} "${${STRING_STRING}}define block last_section with fixed order { };\n")
+        endif()
+        # Not really the right place, we want the last used flash bytes not end of the world!
+        # set(${STRING_STRING} "${${STRING_STRING}}\".last_section\": place at end of ${group_name_lma} { block last_section };\n")
+        set(${STRING_STRING} "${${STRING_STRING}}\".last_section\": place in ${group_name_lma} { block last_section };\n")
```
does this need fixing?

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+
+  set(TEMP "${TEMP}\n{")
+
+  # foreach(start_symbol ${start_syms})
+  #   set(TEMP "${TEMP}\n  section ${start_symbol},")
+  #   set_property(GLOBAL APPEND PROPERTY ILINK_CURRENT_SECTIONS "section ${start_symbol}")
+  # endforeach()
+
+  # if(NOT nosymbols)
+  #   set(TEMP "${TEMP}\n  section __${name_clean}_start,")
+  #   set_property(GLOBAL APPEND PROPERTY ILINK_CURRENT_SECTIONS "section __${name_clean}_start")
+  # endif()
```
is this needed?

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+    elseif(subalign)
+      list(APPEND block_attr "alignment = ${subalign}")
+    elseif(part)
+      # list(APPEND block_attr "alignment = input")
```
here too

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+    endif()
+
+    if(DEFINED symbol_start)
+      # set(TEMP "${TEMP}\n  section ${symbol_start},")
+      # set_property(GLOBAL APPEND PROPERTY ILINK_CURRENT_SECTIONS "section ${symbol_start}")
```
here?

nordicjm commented on cmake/linker/iar/config_file_script.cmake:
```diff
@@ -0,0 +1,922 @@
+      elseif(subalign)
+        list(APPEND block_attr "alignment = ${subalign}")
+      elseif(part)
+        # list(APPEND block_attr "alignment = input")
```
etc.

nordicjm reviewed (dismissed):
build system (non-toolchain) parts OK, comments should be fixed but can be fixed after since experimental

RobinKastberg commented:
Since #84862 just got in, I think I will need to rebase anyway. What happens with a commit in PR that matches one already in tree?
so I can fix the indentation while at it.

tomi-font commented:
> Since #84862 just got in, I think I will need to rebase anyway. What happens with a commit in PR that matches one already in tree? so I can fix the indentation while at it.

It seems like it might be fine without doing anything as the GitHub UI doesn't show that there is a conflict (the commit would just be skipped as it's already present in the tree). Maybe you can confirm by doing a local rebase, and if you don't have a conflict then you don't need to do anything.

kartben reviewed (commented):
Initial docs looks great already, and just tested with a few boards / samples with good success :)
Thanks for the great work!

tejlmand commented on cmake/linker/iar/config_file_script.cmake:
As this is initial experimental IAR support, then I assume more work will be done, and that commented out code is present in this file because this is sensitive areas not yet complete for full IAR support, therefore I have not commented, but please take a second look to see if it really should be commented out or removed because it might be left-over code.

For me and others, a comment describing why code is commented out would also be nice.
(non blocking comment)

tejlmand commented on cmake/linker/iar/target.cmake:
```diff
@@ -0,0 +1,135 @@
+set(ILINK_THUMB_CALLS_WARNING_SUPPRESSED)
+set(IAR_LIB_USED)
+function(toolchain_ld_force_undefined_symbols "")
+#  foreach(symbol ${ARGN})
+#    zephyr_link_libraries(--place_holder=${symbol})
+#  endforeach()
```
What's the plan ?

tejlmand commented on cmake/toolchain/iar/Kconfig:
```diff
@@ -0,0 +1,40 @@
+menu "IAR library options"
+
+config IAR_SEMIHOSTING
+	bool "Use the IAR semihosting implementation."
```
```suggestion
	bool "IAR semihosting implementation."
```
https://docs.zephyrproject.org/latest/build/kconfig/tips.html#prompt-strings
> For a Kconfig symbol that enables a driver/subsystem FOO, consider having just “Foo” as the prompt, instead of “Enable Foo support” or the like.

LoveKarlsson replied:
Ok!

tejlmand commented on lib/libc/Kconfig:
```diff
@@ -51,6 +51,13 @@ config PICOLIBC_SUPPORTED
 
+config IAR_LIBC_SUPPORTED
+	bool
+	default n
```
```suggestion
```

https://docs.zephyrproject.org/latest/build/kconfig/tips.html#redundant-defaults
> bool symbols implicitly default to n, and string symbols implicitly default to the empty string. Therefore, default n and default "" are (almost) always redundant.

tejlmand reviewed (dismissed):
Thanks for the great work.

Did some basic testing on nrf52840dk and the qemu_cortex_m3, and work as expected.
Only observation is that `CONFIG_USE_SEGGER_RTT` must be disabled for linking to succeed, so perhaps for follow-up it would be nice to have this disabled per default, though not a blocker for this PR.

Some minor nits noticed, but not blocking in any way.

:+1:

kartben commented:
just force pushed to fix compliance error. @tejlmand please refresh your +1 :)
https://github.com/zephyrproject-rtos/zephyr/compare/df14d6e52d153b2da64020a1d67df3894a9892ad..f1bf5d3a941b54cd98fcaf6e4cf124a90f7b81cf

LoveKarlsson commented on cmake/compiler/iar/target.cmake:
```diff
@@ -0,0 +1,146 @@
+    --endian=little
+    --cpu=${ICCARM_CPU}
+    -DRTT_USE_ASM=0       #WA for VAAK-232
+    --diag_suppress=Ta184  # Using zero sized arrays except for as last member of a struct is discouraged and dereferencing elements in such an array has undefined behavior
```
I will do better. I usually paste the error message as a comment whenever I suppress an error/warning

ruuddw commented:
See #85854 it looks like this PR broke ARC MetaWare support, and we were not involved in this PR.
Any idea/suggestion what could be wrong?
